In [2]:
#include <iostream>

### Overloading Functions with Different Parameters

In [84]:
int Add(int x, int y){
    return x + y;
};

In [85]:
double Add(double x, double y){
    return x + y;
};

In [86]:
double Add(double x , double y , double z){
    return x + y + z;
};

In [87]:

std::string Add(std::string x, std::string y = " World!"){ // // with default value
    return x + y;
};


In [90]:
std::cout << "Add(1,2) = " << Add(1,2) << std::endl;
std::cout << "Add(1.2,2.3) = " << Add(1.2,2.3) << std::endl;
std::cout << "Add(1.0,2.2,3.5) = " << Add(1.0,2.2,3.5) << std::endl;
std::cout << "Add(\"Hello\") = " << Add("Hello") << std::endl; 
std::cout << "Add(\"Hello\" ,\" Madhawa\") = " << Add("Hello", " Madhawa") << std::endl;

Add(1,2) = 3
Add(1.2,2.3) = 3.5
Add(1.0,2.2,3.5) = 6.7
Add("Hello") = Hello World!
Add("Hello" ," Madhawa") = Hello Madhawa


Function overloading allows multiple functions to have the same name but operate with different types or numbers of inputs. 

### Overloading functions with pointers

In [29]:
void Print(std::string* p_str, size_t count){
    std::cout << "Strings Print : " << std::endl;
    for(int i=0; i < count; i++){
        std::cout << " [" << p_str[i] << "] ";
    }
    std::cout << std::endl;
};

In [30]:
void Print(int* p_int, size_t count){
    std::cout << "Intiger Print : " << std::endl;
    for(int i=0; i < count; i++){
        std::cout << " [" << p_int[i] << "] ";
    }
    std::cout << std::endl;
};

In [31]:
std::string towns[] {"Tokyo","Osaka","Hokkaido","Saitama"};

Print(towns,std::size(towns));

Strings Print : 
 [Tokyo]  [Osaka]  [Hokkaido]  [Saitama] 


In [32]:
int ints[] {1,2,3,4,5,6,7};
Print(ints,std::size(ints));

Intiger Print : 
 [1]  [2]  [3]  [4]  [5]  [6]  [7] 


### Overloading functions with References & Pointers

#### Function 1: Using a Reference

1. Parameter Type: `double& ref_double` takes a reference to a double as a parameter.
2. Function Behavior: Inside the function, `ref_double` directly refers to the variable that was passed in.  
   By adding 2 to `ref_double`, we modify the original variable’s value since the reference is an alias to that variable.

In [ ]:
void AddTwo(double& ref_double){
    std::cout << "as reference: " << std::endl;
    ref_double += 2; // Directly modifies the variable using a reference.
};

In [71]:
double val = 1.2;
AddTwo(val);
std::cout << "new val : " << val << std::endl;

as reference: 
new val : 3.2


* Here, `val` is passed to `AddTwo` by reference.
* The function outputs as reference:, then adds 2 to val, resulting in a new value of 3.2 for val.

#### Function 2: Using a Pointer
1. Parameter Type: `double* p_double` takes a pointer to a double as a parameter.
2. Function Behavior: Inside the function, `p_double[0]` accesses the value at the address to which `p_double` points. Adding 2 to `p_double[0]` modifies the original double value stored at that memory location.

In [ ]:
void AddTwo(double* p_double){ 
    std::cout << "as pointer: " << std::endl;
    p_double[0] += 2; // Modifies the variable by dereferencing the pointer.
};

In [73]:
double val1 = 36.2;
double* p_val1 = &val1;
AddTwo(p_val1);
std::cout << "new val1 : " << val1 << std::endl;

as pointer: 
new val1 : 38.2


* `p_val1` is a pointer to val1, and this pointer is passed to `AddTwo`.
* The function outputs as pointer:, then adds 2 to `val1` through the pointer, resulting in `val1` becoming 38.2.

### More Advances Overloading Consepts

* Overloadibg Operators

In [4]:
class Vector {
public:
    int x, y;
    Vector(int x_val, int y_val) : x(x_val), y(y_val) {}

    Vector operator+(const Vector& other) const {
        return Vector(x + other.x, y + other.y);
    }

    Vector operator-(const Vector& other) const {
        return Vector(x - other.x, y - other.y);
    }
};

In [5]:
Vector v1(3, 4);
Vector v2(1, 2);

Vector result = v1 + v2;
std::cout << "Result of addition: (" << result.x << ", " << result.y << ")\n";

result = v1 - v2;
std::cout << "Result of subtraction: (" << result.x << ", " << result.y << ")\n";

Result of addition: (4, 6)
Result of subtraction: (2, 2)
